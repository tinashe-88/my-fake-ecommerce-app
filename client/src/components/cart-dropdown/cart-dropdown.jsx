import React from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { createStructuredSelector } from 'reselect'

import CartItem from '../cart-item/cart-item'

import  { selectCartItems } from '../../redux/cart/cart.selectors'
import { toggleCartHidden } from '../../redux/cart/cart.actions'

import { 
  CartDropdownContainer,
  CartDropdownButton,
  CartItemsContainer,
  EmptyMessageContainer
} from './cart-dropdown.styles'

export const CartDropdown = ({ cartItems, history, dispatch }) => (
  <CartDropdownContainer>
    <CartItemsContainer>
      {
        cartItems.length ?
        cartItems.map(cartItem => (
          <CartItem key={cartItem.id} item={cartItem} />
        )) 
        : (
          <EmptyMessageContainer>
            Your cart is empty
          </EmptyMessageContainer>
        )
      }
    </CartItemsContainer>
    <CartDropdownButton 
      onClick={() => {
        history.push('/checkout')
        dispatch(toggleCartHidden())     
      }}
    >
      Go To Checkout
    </CartDropdownButton>
  </CartDropdownContainer>
)

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems
})

export default withRouter(connect(mapStateToProps)(CartDropdown))