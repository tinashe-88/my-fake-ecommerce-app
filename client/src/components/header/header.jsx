import React from 'react'
import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'

import CartIcon from '../cart-icon/cart-icon'
import CartDropdown from '../cart-dropdown/cart-dropdown'

import { selectCartHidden } from '../../redux/cart/cart.selectors'
import { selectCurrentUser } from '../../redux/user/user.selector'
import { signOutStart } from '../../redux/user/user.actions'

import {
  HeaderContainer,
  OptionsContainer,
  OptionLink,
  LogoContainer,
} from './header.styles'

import {
  ReactComponent as Logo
} from '../../assets/fake-logo.svg'

export const Header = ({ currentUser, hidden, signOutStart }) => (
  <HeaderContainer>
    <LogoContainer
      to='/'
    >
      <Logo />
    </LogoContainer>
    <OptionsContainer>
      <OptionLink 
        to='/shop'
      >
        Shop
      </OptionLink>
      <OptionLink 
        to='/contact'
      >
        Contact
      </OptionLink>
      {
        currentUser ? (
          <OptionLink
            as='div'
            onClick={signOutStart}
          >
            Sign Out
          </OptionLink>
        ) : (
          <OptionLink 
            to='/signin'
          >
            Sign In
          </OptionLink>
        )  
      }
      <CartIcon />
    </OptionsContainer>
    {
      hidden ? null : <CartDropdown/>
    }
  </HeaderContainer>
)

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden
})

const mapDispatchToProps = dispatch => ({
  signOutStart: () => dispatch(signOutStart())
})

export default connect(
  mapStateToProps, 
  mapDispatchToProps
)(Header)