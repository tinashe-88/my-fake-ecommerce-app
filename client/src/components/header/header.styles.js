import styled, { css } from 'styled-components'
import { Link } from 'react-router-dom'

const OptionContainerStyles = css`
  color: #000;
  padding: 10px 15px;
  cursor: pointer; 
`

const HoverLinks = css`
  color: #ffa3d7;
  transition: color .15s ease-in;
`

export const HeaderContainer = styled.div`
  height: 70px;
  width: 100%;
  display: flex;
  justify-content: space-between;
  margin-bottom: 25px;
  border-bottom: 1px solid #f4f4f4;

  @media screen and (max-width: 800px) {
    height: 60px;
    padding: 10px;
    margin-bottom: 20px;
  }
`

export const LogoContainer = styled(Link)`
  height: 100%;
  width: 70px;
  padding-top: 11px;
  color: #fff;

  @media screen and (max-width: 800px) {
    width: 50px;
    padding: 0;
  }
`

export const OptionsContainer = styled.div`
  width: 50%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: space-around;

  @media screen and (max-width: 800px) {
    width: 80%; 
  }
`

export const OptionLink = styled(Link)`
  ${OptionContainerStyles}

  &:hover {
    ${HoverLinks}
  }

  @media screen and (max-width: 500px) {
    font-size: 12pt;
  }
`

OptionLink.displayName = 'OptionLink'

export const OptionDiv = styled.div`
  ${OptionContainerStyles}

  &:hover {
      ${HoverLinks}
  }
`