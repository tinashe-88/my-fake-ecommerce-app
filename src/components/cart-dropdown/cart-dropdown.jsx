import React from 'react'
import { connect } from 'react-redux'

import  { selectCartItems } from '../../redux/cart/cart.selectors'

import CartItem from '../cart-item/cart-item'

import { 
  CartDropdownContainer,
  CartDropdownButton,
  CartItemsContainer,
  EmptyMessageContainer
} from './cart-dropdown.styles'

const CartDropdown = ({ cartItems }) => (
  <CartDropdownContainer>
    <CartItemsContainer>
    {cartItems.map(cartItem => (
      <CartItem key={cartItem.id} item={cartItem} />
    ))}
    </CartItemsContainer>
    <CartDropdownButton>Go to checkout</CartDropdownButton>
  </CartDropdownContainer>
)

const mapStateToProps = state => ({
  cartItems: selectCartItems(state)
})

export default connect(mapStateToProps)(CartDropdown)