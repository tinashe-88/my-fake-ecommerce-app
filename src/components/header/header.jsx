import React from 'react'

import { Link } from 'react-router-dom'

import {
  ReactComponent as Logo
} from '../../assets/fake-logo.svg'

import {
  HeaderContainer,
  OptionsContainer
} from './header.styles'

const Header = () => (
  <HeaderContainer>
    <Link 
      to='/'
      className='logo-container'
    >
      <Logo />
    </Link>
    <OptionsContainer>
      <Link 
        to='/shop'
        className='option'
      >
        Shop
      </Link>
      <Link 
        to='/contact'
        className='option'
      >
        Contact
      </Link>
    </OptionsContainer>
  </HeaderContainer>
)

export default Header