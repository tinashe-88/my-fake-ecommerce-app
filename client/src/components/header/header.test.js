import React from 'react'
import { shallow } from 'enzyme'

import { Header } from './header'
import { CartDropdown } from '../cart-dropdown/cart-dropdown.jsx'

describe('Header component', () => {
  let wrapper
  let mockSignOutStart

  beforeEach(() => {
    mockSignOutStart = jest.fn()

    const mockProps = {
      hidden: true,
      currentUser: {
        uid: '332'
      },
      signOutStart: mockSignOutStart
    }

    wrapper = shallow(<Header {...mockProps}/>)
  })

  it('Should render Header component', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('If currentUser is present', () => {
    it('Should render signout link', () => {
      expect(
        wrapper.find('OptionLink').at(2).text()
      ).toBe('Sign Out')
    })

    it('Should call signOutStart method when link is clicked', () => {
      wrapper.find('OptionLink').at(2).simulate('click')
      expect(mockSignOutStart).toHaveBeenCalled()
    })
  })
})