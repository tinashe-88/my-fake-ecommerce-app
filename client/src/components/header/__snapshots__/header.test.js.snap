// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Should render Header component 1`] = `
<styled.div>
  <Styled(Link)
    to="/"
  >
    <ForwardRef(SvgFakeLogo) />
  </Styled(Link)>
  <styled.div>
    <OptionLink
      to="/shop"
    >
      Shop
    </OptionLink>
    <OptionLink
      to="/contact"
    >
      Contact
    </OptionLink>
    <OptionLink
      as="div"
      onClick={[MockFunction]}
    >
      Sign Out
    </OptionLink>
    <Connect(CartIcon) />
  </styled.div>
</styled.div>
`;
